chaincode/marbles_transfer/marbles_transfer.go

```diff
@@ -0,0 +1,417 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package main
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+
+	"github.com/hyperledger/fabric-chaincode-go/pkg/statebased"
+	"github.com/hyperledger/fabric-chaincode-go/shim"
+	"github.com/hyperledger/fabric-contract-api-go/contractapi"
+)
+
+const (
+	typeMarbleForSale = "S"
+	typeMarbleBid     = "B"
+)
+
+type SmartContract struct {
+	contractapi.Contract
+}
+
+// Marble struct and properties must be exported (start with capitals) to work with contract api metadata
+type Marble struct {
+	ObjectType        string `json:"object_type"` // ObjectType is used to distinguish different object types in the same chaincode namespace
+	ID                string `json:"marble_id"`
+	OwnerOrg          string `json:"owner_org"`
+	PublicDescription string `json:"public_description"`
+}
+
+// CreateMarble creates a marble and sets it as owned by the client's org
+func (s *SmartContract) CreateMarble(ctx contractapi.TransactionContextInterface, marbleID string) error {
+
+	transMap, err := ctx.GetStub().GetTransient()
+	if err != nil {
+		return fmt.Errorf("Error getting transient: " + err.Error())
+	}
+
+	// Marble properties are private, therefore they get passed in transient field
+	immutablePropertiesJSON, ok := transMap["marble_properties"]
+	if !ok {
+		return fmt.Errorf("marble_properties key not found in the transient map")
+	}
+
+	// Get client org id and verify it matches peer org id.
+	// In this scenario, client is only authorized to read/write private data from its own peer.
+	clientOrgID, err := getClientOrgID(ctx, true)
+	if err != nil {
+		return fmt.Errorf("failed to get verified OrgID: %s", err.Error())
+	}
+
+	// Create and persit marble
+
+	marble := Marble{
+		ObjectType:        "marble",
+		ID:                marbleID,
+		OwnerOrg:          clientOrgID,
+		PublicDescription: "A new marble for " + clientOrgID,
+	}
+
+	marbleJSON, err := json.Marshal(marble)
+	if err != nil {
+		return fmt.Errorf("failed to create marble JSON: %s", err.Error())
+	}
+
+	err = ctx.GetStub().PutState(marble.ID, marbleJSON)
+	if err != nil {
+		return fmt.Errorf("failed to put Marble in public data: %s", err.Error())
+	}
+
+	// Set the endorsement policy such that an owner org peer is required to endorse future updates
+	err = setMarbleStateBasedEndorsement(ctx, marble.ID, clientOrgID)
+	if err != nil {
+		return fmt.Errorf("failed setting state based endorsement for owner: %s", err.Error())
+	}
+
+	// Persist private immutable marble properties to owner's private data collection
+	collection := "_implicit_org_" + clientOrgID
```

Comment:
Not your fault, but this `_implicit_org_` thing looks like a pretty ugly hack. I wonder if we should make a tactical fix to the stub APIs to add a `GetImplicitCollectionName` method so we can hide it behind the scenes where it belongs?